### PR TITLE
Add ;;weather command

### PR DESCRIPTION
**Pull request by SuhanKoh**
**Sample output on the text channel**

``;;weather san francsico``
weather1

``;;weather singapore``
weather2

### CODE REVIEW

File: FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java [hunk 1]
```diff
@@ -84,6 +82,7 @@ public static void initCommands() {
         CommandRegistry.registerCommand("export", new ExportCommand(), "ex");
         CommandRegistry.registerCommand("gr", new GensokyoRadioCommand(), "gensokyo", "gensokyoradio");
         CommandRegistry.registerCommand("muserinfo", new UserInfoCommand());
+        CommandRegistry.registerCommand("weather", new WeatherCommand(new OpenWeatherAPI()));
```
Comment (Frederikam) [inline FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: This is not a music command

Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: Woops. Fixing.

File: FredBoat/src/main/java/fredboat/util/rest/models/weather/RetrievedWeather.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * Interface for getting weather data.
+ */
+public interface RetrievedWeather {
+
+    /**
+     * Error indication for retrieving weather.
+     *
+     * @return True if there is an error, false if successful.
+     */
+    boolean IsError();
```
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/util/rest/models/weather/RetrievedWeather.java]: Could you name this `isError()`?

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+public class WeatherCommand extends Command implements IUtilCommand {
+
+    private Weather weather;
+
+    public WeatherCommand(Weather weatherImplementation) {
+        weather = weatherImplementation;
+    }
+
+    @Override
+    public void onInvoke(Guild guild, TextChannel channel, Member invoker, Message message, String[] args) {
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: When you rebase onto development, you will notice that the signature of commands has changed. Here's what changed: Calls to `Command#onInvoke` get a CommandContext parameter now that provides all other parameters, and should also be used to send any messages via `CommandContext#reply` (or any other messaging convenience methods it provides).

File: FredBoat/src/main/java/fredboat/util/rest/models/weather/RetrievedWeather.java (hunk 2 above)
Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/util/rest/models/weather/RetrievedWeather.java]: Yep 👍

**Comment by napstr** [comment]
Hi, this is missing a PR against https://github.com/Frederikam/FredBoat-i13n with the new i18n strings used in here.
You only need to add them to the en_US file, we will then upload them to crowdin and crowdin will update the other language files.

**Comment by SuhanKoh** [comment]
@napstr Thanks, working on it

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                            MessageFormat.format(I18n.get(context.guild).getString("weatherTemperature"),
+                                    currentWeather.getTemperature()));
+
+                    CentralMessaging.editMessage(outMsg, messageBuilder.build());
+                }
+                return;
+            }
+
+            CentralMessaging.editMessage(outMsg,
+                    MessageFormat.format(I18n.get(context.guild).getString("weatherUsageError"),
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: This is a good place to use `HelpCommand.sendFormattedCommandHelp(context)` instead of a special error, since the help should contain the exact usage of a command and arguments.

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                            Config.CONFIG.getPrefix()));
+        });
+    }
+
+    @Override
+    public String help(Guild guild) {
+        return I18n.get(guild).getString("helpWeatherCommand");
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: This string is missing in the i13n PR.

**Comment by SuhanKoh** [comment]
- Rework result output to provide embedded picture
- Fix i13n strings

**Comment by SuhanKoh** [comment]
@napstr 
Preview
image

Should I add anything else?

**Comment by napstr** [comment]
Looks good to me.
If you are going to stick with this you could actually scrape the "Current weather is" string since that's exactly what the user just asked for.

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java (hunk 4 above)
Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: Ah I missed this comment, fixing it now.

**Comment by napstr** [comment]
To merge this, this needs to be rebased onto the CommandContext changes at minimum.

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    private static final String HELP_STRING_FORMAT = "{0}{1} <location>\n#";
+
+    public WeatherCommand(Weather weatherImplementation) {
+        weather = weatherImplementation;
+    }
+
+    @Override
+    public void onInvoke(CommandContext context) {
+
+        context.reply(I18n.get(context.guild).getString("weatherLoading"), outMsg -> {
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: This use of queue means we aren't doing the call to the Weather API or anything else until discord gets back to us that our message has been posted. This makes our response artificially and unnecessarily slow.

General note on the Discord calls the WeatherCommand is making:
Status quo:
1. Announce the fetching
2. Delete the fetching message
3. Post the Embed

Instead of emulating such a "typing" event with the fetch message, I suggest to use `CommandContext#sendTyping`, while querying the WeatherAPI, and then post the embed with results as soon as we receive it (or an error message if it anything wrong).
That slims it down to 2 REST actions instead of 3 per weather command call.

**Comment by napstr** [comment]
Addendum, github somehow ate this one:

 - How are the OpenWeather ratelimits handled?
They have a rather harsh punishment on [hitting their ratelimit](https://openweathermap.org/appid), don't want that to happen:

> Access limitation
If account exceeds the limits, then a notification about limits exceeding is sent. If it repeats again, then the account is blocked for an hour. Therefore, the lock period is increased by one hour until 4 hours block sets. When blocking repeats the fifth time, then the lock period lasts 24 hours. This rule is cycled. Please be carefull with the number of API calls you complete. 

 - We check other APIs, like imgur or MAL, at start up. You can see that code in `FredBoat.java`, please add a similar thing for the Open Weather API.

- Also, I got an unexpected error, but there was nothing logged:
image
While the log looked like this (I know the times arent the same, but it really looked like that):
> [16:04:49] [ INFO ] [EventListenerBoat] secrit debag servur 	 fry 	 <<weather Hamburg
> [16:04:49] [ INFO ] [EventListenerBoat] secrit debag servur 	 FredBoatTest 	 Fetching weather...

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    public WeatherCommand(Weather weatherImplementation) {
+        weather = weatherImplementation;
+    }
+
+    @Override
+    public void onInvoke(CommandContext context) {
+
+        context.sendTyping();
+        MessageFuture future = context.reply(I18n.get(context.guild).getString("weatherLoading"));
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: There is no need to do a typing event AND a message. Sending a message cancels the display of the typing event in the Discord client anyways. Please just do the typing event.

File: FredBoat/src/main/java/fredboat/FredBoat.java [hunk 8]
```diff
@@ -279,6 +283,22 @@ private static boolean hasValidImgurCredentials() {
         return false;
     }
 
+    /**
+     * Method to check if there is an error to retrieve open weather data.
+     *
+     * @return True if it can retrieve data, else return false.
+     */
+    private static boolean hasValidOpenWeatherKey() {
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/FredBoat.java]: Is there are reason that this method, unlike the other API checks, is missing logging of its success / failure?

File: FredBoat/src/main/java/fredboat/util/rest/OpenWeatherAPI.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        client = new OkHttpClient();
+        objectMapper = new ObjectMapper();
+
+        currentWeatherBaseUrl = HttpUrl.parse(OPEN_WEATHER_BASE_URL + "/weather");
+        if (currentWeatherBaseUrl == null) {
+            log.debug("Open weather search unable to build URL");
+        }
+    }
+
+    public RetrievedWeather getCurrentWeatherByCity(@NotNull String query) {
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/util/rest/OpenWeatherAPI.java]: Please use the `javax.annotation.Nonnull` annotation (in all the places too).

Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/util/rest/OpenWeatherAPI.java]: Got it.

File: FredBoat/src/main/java/fredboat/FredBoat.java (hunk 8 above)
Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/FredBoat.java]: Correct, they do not provide api to check, it will just return status code 439. Which is already logged in the API implementation.

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java (hunk 7 above)
Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: Ok

File: FredBoat/src/main/java/fredboat/FredBoat.java (hunk 8 above)
Comment (napstr) [reply FredBoat/src/main/java/fredboat/FredBoat.java]: 439 seems like a weird code to provide.
The point of doing the check here is to have comprehensible output of whats happening, here's an example:
```
log.warn("Open Weather API credentials not found. Weather related commands will not work properly.");
log.info("Open Weather API check successful");
log.warn("Open Weather API check failed. It may be down, the provided credentials may be invalid, or temporarily blocked.", e);
```

**Comment by napstr** [comment]
Btw, here's the full representation of how their rate limit answer looks (for science). Status code is 429.
```
{
	"cod": 429,
	"message": "Your account is temporary blocked due to exceeding of requests limitation of your subscription type. Please choose the proper subscription http://openweathermap.org/price"
}

Server: openresty
Date: Sat, 30 Sep 2017 07:51:10 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 196
Connection: keep-alive
X-Cache-Key: /data/2.5/weather?APPID=SNIP&q=hamburg
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET, POST
```

**Comment by napstr** [comment]
Hey, I fear I my explanation of things might have been ambiguous with regards to the usage of the word "ratelimit".

Let's try to clear things up:
The [FredBoat Ratelimiter](https://github.com/Frederikam/FredBoat/blob/development/FredBoat/src/main/java/fredboat/util/ratelimit/Ratelimiter.java) is a tool to limit command spam from users. It does not and can not enforce and respect the rate limits of another service - this is exactly the opposite of that. Users have to respect the FredBoat ratelimiter, just as we have to respect rate limits of the various APIs we use.
The important part here is, that it is mostly just a tool to reduce load on our own infrastructure and the services behind that.

Respecting the ratelimit of the Open Weather API requires you to write some custom code and it is really up to you how to do that. A good starting point is reading about Token Buckets and Leaky Buckets algorithms, the latter being a better fit in this case imo.


Speaking of the FredBoat ratelimiter, I see that you added a limit of 2 calls per 2 minutes on a guild scope, which I find to be too harsh. The goal of the FredBoat ratelimiter is to be never seen by regular users. Here's what I had in mind: 10 calls / 3 minutes per user, 30 calls / 3 minutes per guild.


To sum things up, this PR looks pretty much done and I would love to merge it. The only two things missing are respecting the Open Weather API ratelimit and caching requests. I can understand that these things might be out of scope for your initial plans writing this PR, so I am willing to accept the PR without them, and assign myself (if anyone reading this wants to do it feel free to speak up) to add those two missing things. Please let me know if that's ok for you, or if you want to try adding those things yourself.

**Comment by SuhanKoh** [comment]
- Add api limiter
- Change the ratelimiter to 30 calls / 3 minutes

**Comment by napstr** [comment]
https://ratelimits.are-la.me/4914df.png
```
[13:13:07] [ ERROR] [TextUtils] Caught exception while executing a command
com.google.common.util.concurrent.UncheckedExecutionException: fredboat.util.rest.APILimitException: OpenWeather API maximum rate exceeded.
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2218)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4147)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4151)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:5140)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:5146)
	at fredboat.util.rest.OpenWeatherAPI.getCurrentWeatherByCity(OpenWeatherAPI.java:72)
	at fredboat.command.util.WeatherCommand.onInvoke(WeatherCommand.java:42)
	at fredboat.commandmeta.CommandManager.prefixCalled(CommandManager.java:135)
	at fredboat.event.EventListenerBoat.limitOrExecuteCommand(EventListenerBoat.java:143)
	at fredboat.event.EventListenerBoat.onMessageReceived(EventListenerBoat.java:119)
	at net.dv8tion.jda.core.hooks.ListenerAdapter.onEvent(ListenerAdapter.java:404)
	at net.dv8tion.jda.core.hooks.InterfacedEventManager.handle(InterfacedEventManager.java:84)
	at net.dv8tion.jda.core.handle.MessageCreateHandler.handleDefaultMessage(MessageCreateHandler.java:128)
	at net.dv8tion.jda.core.handle.MessageCreateHandler.handleInternally(MessageCreateHandler.java:49)
	at net.dv8tion.jda.core.handle.SocketHandler.handle(SocketHandler.java:37)
	at net.dv8tion.jda.core.requests.WebSocketClient.handleEvent(WebSocketClient.java:969)
	at net.dv8tion.jda.core.requests.WebSocketClient.onTextMessage(WebSocketClient.java:661)
	at com.neovisionaries.ws.client.ListenerManager.callOnTextMessage(ListenerManager.java:352)
	at com.neovisionaries.ws.client.ReadingThread.callOnTextMessage(ReadingThread.java:260)
	at com.neovisionaries.ws.client.ReadingThread.callOnTextMessage(ReadingThread.java:238)
	at com.neovisionaries.ws.client.ReadingThread.handleTextFrame(ReadingThread.java:963)
	at com.neovisionaries.ws.client.ReadingThread.handleFrame(ReadingThread.java:746)
	at com.neovisionaries.ws.client.ReadingThread.main(ReadingThread.java:108)
	at com.neovisionaries.ws.client.ReadingThread.runMain(ReadingThread.java:64)
	at com.neovisionaries.ws.client.WebSocketThread.run(WebSocketThread.java:45)
Caused by: fredboat.util.rest.APILimitException: OpenWeather API maximum rate exceeded.
	at fredboat.util.rest.OpenWeatherAPI.processGetWeatherByCity(OpenWeatherAPI.java:127)
	at fredboat.util.rest.OpenWeatherAPI.access$000(OpenWeatherAPI.java:28)
	at fredboat.util.rest.OpenWeatherAPI$1.load(OpenWeatherAPI.java:58)
	at fredboat.util.rest.OpenWeatherAPI$1.load(OpenWeatherAPI.java:55)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3708)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2416)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2299)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2212)
	... 24 common frames omitted
```

Suggestion: Using or extending FredBoats MessagingException which is rendered in a much nicer way for the user.

**Comment by Shredder121** [comment]
My comment was addressed, basically about naming convention.
Napster's got this. 👍

**Comment by napstr** [comment]
Hey, summary of what we just talked about in Discord & more:

- When the user enters a query that is entirely non-alphanumeric we shouldn't send it. This is what we are currently getting as a result if we try:
`[21:40:07] [ WARN ] [OpenWeatherAPI] {"cod":"400","message":"Nothing to geocode"}`
- When a query is send that doesn't return a result we are logging that as a warning which means it will end up in sentry. That is not acceptable, as it is due to wrong user input.
`[21:49:00] [ WARN ] [OpenWeatherAPI] {"cod":"404","message":"city not found"}`
- While you're at it, please add an @Override annotation to the getCurrentWeatherByCity() method in OpenWeatherApi
- The Cache eats our own MessagingExceptions and throws them as ExecutionExceptions/UncheckedExecutionExceptions (see my post above for an example trace). Here's is an ugly hack that works:tm: around that so that we can use our own MessagingExceptions:
```
        try {
            return weatherCache.getUnchecked(query);
        } catch (UncheckedExecutionException e) {
            Throwable cause = e.getCause();
            if (cause instanceof MessagingException) {
                throw (MessagingException) cause;
            } else {
                throw e;
            }
        }
```
Feel free to use something else if you find a more elegant way.

File: FredBoat/src/main/java/fredboat/commandmeta/init/MainCommandInitializer.java [hunk 10]
```diff
@@ -79,6 +80,8 @@ public static void initCommands() {
         CommandRegistry.registerCommand("catgirl", new CatgirlCommand(), "neko", "catgrill");
         CommandRegistry.registerCommand("avatar", new AvatarCommand(), "ava");
         CommandRegistry.registerCommand("say", new SayCommand());
+        // TODO: Need to implement api limit before using this, OpenWeather has a harsh api limit.
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/commandmeta/init/MainCommandInitializer.java]: This todo looks done to me.

File: FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java [hunk 11]
```diff
@@ -39,13 +39,11 @@
...
 import fredboat.command.music.seeking.SeekCommand;
-import fredboat.command.util.CommandsCommand;
-import fredboat.command.util.HelpCommand;
-import fredboat.command.util.MusicHelpCommand;
-import fredboat.command.util.UserInfoCommand;
+import fredboat.command.util.*;
 import fredboat.commandmeta.CommandRegistry;
 import fredboat.perms.PermissionLevel;
 import fredboat.shared.constant.DistributionEnum;
+import fredboat.util.rest.OpenWeatherAPI;
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: Please don't reformat this file.

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                    if (currentWeather.getThumbnailUrl().length() > 0) {
+                        embedBuilder.setThumbnail(currentWeather.getThumbnailUrl());
+                    }
+
+                    context.reply(embedBuilder.build());
+
+                } else {
+                    switch (currentWeather.errorType()) {
+                        case LOCATION_NOT_FOUND:
+                            context.reply(MessageFormat.format(I18n.get(context.guild).getString("weatherLocationNotFound"), query));
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: Please use the Context#i18nFormat method.

Please put the user input into code brackets \`\`. Here's how it would look:
image

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                            break;
+
+                        default:
+                            context.reply(
+                                    MessageFormat.format(I18n.get(context.guild).getString("weatherError"),
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: Please use the Context#i18nFormat method.

Please put the user input into code brackets \`\`. Here's how it would look (this is another message, but you get the point):
image

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+                                            query.toUpperCase()));
+                            break;
+                    }
+                }
+                return;
+
+            } catch (APILimitException e) {
+                context.reply(I18n.get(context.guild).getString("tryLater"));
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: Please use the Context#i18n method.

File: FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        }
+
+        sendHelpString(context);
+    }
+
+    @Nonnull
+    @Override
+    public String help(@Nonnull Context context) {
+        return HELP_STRING_FORMAT + I18n.get(context.getGuild()).getString("helpWeatherCommand");
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/util/WeatherCommand.java]: Please use the Context#i18n method.

File: FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java (hunk 11 above)
Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: the util.* import seems to be imported via https://github.com/Frederikam/FredBoat/blob/development/FredBoat/src/main/java/fredboat/commandmeta/init/MainCommandInitializer.java

**Comment by SuhanKoh** [comment]
- [ ] Remove TODO
- [ ] Use Context#18n

File: FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java (hunk 11 above)
Comment (Shredder121) [reply FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: We're saying the music command handler hasn't been touched apart from a reformat.
So then don't touch it at all. :(

Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: Oh yeah, this line isn't supposed to be here.

File: FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java [hunk 16]
```diff
@@ -39,10 +39,7 @@
 import fredboat.command.music.seeking.RestartCommand;
 import fredboat.command.music.seeking.RewindCommand;
 import fredboat.command.music.seeking.SeekCommand;
-import fredboat.command.util.CommandsCommand;
-import fredboat.command.util.HelpCommand;
-import fredboat.command.util.MusicHelpCommand;
-import fredboat.command.util.UserInfoCommand;
```
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: These are still touched.
If in the end nothing meaningful in the file is changed then it's best to undo any changes in the file so it's not a part of the diff anymore that can cause conflicts down the line.

**Comment by napstr** [comment]
Wew. guess what, a few more things : ^)

1. Address this please. https://github.com/Frederikam/FredBoat/pull/290#discussion_r143326377

2. From their site: https://openweathermap.org/price

> You can freely use our products and data with non-commercial or commercial purposes. However, the OpenWeatherMap should be credited as a weather data source in a visible part of your application.

This is probably best solved by using the EmbedBuilder#setAuthor or EmbedBuilder#setFooter methods providing credit to the source of the weather.

3. Please set the color of the Embed to the FredBoat color found in BotConstants.java

**Comment by napstr** [comment]
I think this would be an acceptable way to do it:
image

If you are going to use their logo, please use this link: https://i.imgur.com/YqZuqEB.jpg

File: FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java (hunk 16 above)
Comment (SuhanKoh) [reply FredBoat/src/main/java/fredboat/commandmeta/init/MusicCommandInitializer.java]: ok

**Comment by napstr** [comment]
Now both CommandInitializer classes have fucked up formatting. The color for the Embed is also missing.
